Throttle topic administrative functions

graphcareful opened the pull request:
This PR implements [KIP-599](https://cwiki.apache.org/confluence/display/KAFKA/KIP-599%3A+Throttle+Create+Topic%2C+Create+Partition+and+Delete+Topic+Operations) Throttle topic administrative functions, namely they are, `CreateTopics`, `DeleteTopics` and `CreatePartitions` requests.

Importantly to note, this KIP doesn't attempt to constrain the number of global partition mutations, cluster wide. The KIP enforces the quota at the client level. We could have deviated from the KIP in this respect, but there is already ongoing work and planning within the multi-tenancy effort to enforce such limits, so we thought it was best to implement the KIP as it originally stands.

Throttling works by using a modified token-bucket algorithm where the `tokens` are representative to the number of partition mutations made by one of the three aforementioned requests. The algorithm is modified to accept a given `burst` which allows a burst of several mutations with a small or medium number of partitions within the same request to proceed. Subsequent requests will observe an exceeded quota and will return a newly implemented error code `THROTTLING_QUOTA_EXCEEDED` back to the client (if the client is new enough) along with the total throttle ms time in the response.

jcsp commented on src/v/kafka/server/requests.cc:
```diff
@@ -225,6 +225,19 @@ bool track_latency(api_key key) {
+// partitions requests
+bool track_partition_mutations(api_key key) {
+    switch (key) {
+    case create_topics_handler::api::key:
```
I know this is mirroring the KIP-599 list, but I wonder if we should take the opportunity to define a broader definition.  For example, we could include topic config changes.

It would be comforting to know that we had a relatively broad throttle on Kafka operations that can result in controller log appends, because we currently don't compact the controller log, and rely on the overall number of entries being "reasonable"

graphcareful replied:
Interesting, yeah we don't have to strictly adhere to the KIP at all. Are topic config changes equally as expensive as partition mutations?

jcsp replied:
Not quite equally (creating partitions is inherently more expensive in terms of going out there and creating log segments etc), but they're both things that can cause serious problems if we let users flood us.

Maybe we should apply the rate limit to anything that's a controller log write?  That would include things like users and acls.  What do you think @mmaslankaprv ?

graphcareful replied:
FWIW Kafka addresses this by tracking different quotas and having a manager for each, due to the fact that not all requests are equally as expensive for the controller to handle.

jcsp commented:
Missing release notes section in PR description (should describe the feature and specify the new config property name)

jcsp commented on src/v/cluster/errc.h:
```diff
@@ -59,6 +59,7 @@ enum class errc : int16_t {
     error_collecting_health_report,
     leadership_changed,
     feature_disabled,
+    quota_exceeded,
```
It's odd to define this in the cluster namespace when the enforcement is all in code in the kafka namespace

graphcareful replied:
Old code, nice catch

jcsp commented on src/v/cluster/errc.h:
```diff
@@ -59,6 +59,7 @@ enum class errc : int16_t {
     error_collecting_health_report,
     leadership_changed,
     feature_disabled,
+    quota_exceeded,
```
I don't see `quota_exceeded` being returned anywhere?  It's referenced in the path for converting to a kafka exception, but I don't see any code that actually originates an error of this type.

jcsp commented on src/v/kafka/protocol/errors.h:
```diff
@@ -220,6 +220,8 @@ enum class error_code : int16_t {
     // This record has failed the validation on broker and hence be rejected.
     invalid_record = 87,
+    // Broker declined to process request due to exceeded resource quotas.
+    throttling_quota_exceeded = 89,
```
The KIP mentions doing client version checks before sending this new-ish response code  -- do we need to do that?

(although per my comment on cluster::errc::quota_exceeded, I'm not sure if we're actually ever returning this error)

graphcareful replied:
Fixed

jcsp commented:
> Importantly to note, this KIP doesn't attempt to constrain the number of global partition mutations, cluster wide. The KIP enforces the quota at the client level. 

The KIP enforces on a per-node+per-client basis, right?  Whereas this PR is enforcing on a per-shard+per-client basis, which is quite a bit looser.  From a superficial reading, I think this only protects us from aggressive clients if the client is using a single connection: if they opened a bunch of connections (that might hit different shards), the per-client limits are less meaningful.

Per-shard enforcement of quotas is fine if the resource we're protecting is also per-shard (cpu time, memory), but in this instance if we're trying to protect our central controller log, I'm not sure this is going to give us as much confidence as we would like.  Maybe the enforcement should be done on the controller leader instead?  Or perhaps even in both places: a "friendly" check at the kafka level that does this sleep-based throttling, plus a "hard" check on the controller leader that would actually send the quota_exceeded errors if the controller log is being written to too fast.

BenPope commented on src/v/kafka/server/handlers/create_partitions.cc:
```diff
@@ -208,6 +208,23 @@ ss::future<response_ptr> create_partitions_handler::handle(
+    const auto mutations = std::accumulate(
+      request.data.topics.begin(),
+      valid_range_end,
+      static_cast<std::size_t>(0),
```
nit, you can construct a `size_t` directly: `std::size_t{0}`

BenPope commented on src/v/kafka/server/handlers/create_topics.cc:
```diff
@@ -166,6 +167,18 @@ ss::future<response_ptr> create_topics_handler::handle(
+          const auto mutations = std::accumulate(
+            request.data.topics.begin(),
+            request.data.topics.end(),
+            static_cast<std::size_t>(0),
```
`std::size_t{0}`

BenPope commented on src/v/kafka/server/handlers/delete_topics.cc:
```diff
@@ -82,6 +84,19 @@ delete_topics_handler::handle(request_context ctx, ss::smp_service_group) {
+        const auto mutations = std::accumulate(
+          request.data.topic_names.begin(),
+          request.data.topic_names.end(),
+          static_cast<std::size_t>(0),
```
nit: `std::size_t{0}`

BenPope commented on src/v/kafka/server/handlers/delete_topics.cc:
```diff
@@ -82,6 +84,19 @@ delete_topics_handler::handle(request_context ctx, ss::smp_service_group) {
+          static_cast<std::size_t>(0),
+          [&ctx](std::size_t acc, const model::topic& t) {
+              const auto cfg = ctx.metadata_cache().get_topic_cfg(
+                model::topic_namespace(model::kafka_namespace, t));
```
`topic_namespace_view` is sufficient here.

BenPope commented on src/v/kafka/server/handlers/create_partitions.cc:
```diff
@@ -208,6 +208,23 @@ ss::future<response_ptr> create_partitions_handler::handle(
+      static_cast<std::size_t>(0),
+      [&ctx](std::size_t acc, const kafka::create_partitions_topic& t) {
+          const auto cfg = ctx.metadata_cache().get_topic_cfg(
+            model::topic_namespace(model::kafka_namespace, t.name));
```
`topic_namespace_view` is sufficient.

BenPope commented on src/v/kafka/server/handlers/create_partitions.cc:
```diff
@@ -161,6 +161,19 @@ ss::future<response_ptr> create_partitions_handler::handle(
+      "Partition count must be greater then current number of partitions",
+      [&ctx](const create_partitions_topic& tp) {
+          return tp.count > ctx.metadata_cache()
+                              .get_topic_cfg(model::topic_namespace(
```
`topic_namespace_view` should suffice

graphcareful commented on src/v/config/configuration.cc:
```diff
@@ -226,6 +226,14 @@ configuration::configuration()
+  , partition_mutation_rate(
+      *this,
+      "partition_mutation_rate",
+      "Target quota rate (partition mutations per default_window_sec) - 5k "
```
Nice catch, old code, this is an optional bounded property and when its set to nullopt the idea is the feature is disabled and theres no rate limiting, kafka works the same way

graphcareful commented on src/v/config/configuration.cc:
```diff
@@ -226,6 +226,14 @@ configuration::configuration()
       {.min = 1_MiB})
+  , partition_mutation_rate(
+      *this,
+      "partition_mutation_rate",
```
Kafka calls it `controller_mutations_rate`, IMO `admin_rate_limit` might make people think that we are throttling the entire kafka admin API which this PR does not do

graphcareful commented:
> > Importantly to note, this KIP doesn't attempt to constrain the number of global partition mutations, cluster wide. The KIP enforces the quota at the client level.
> 
> The KIP enforces on a per-node+per-client basis, right? Whereas this PR is enforcing on a per-shard+per-client basis, which is quite a bit looser. From a superficial reading, I think this only protects us from aggressive clients if the client is using a single connection: if they opened a bunch of connections (that might hit different shards), the per-client limits are less meaningful.
> 
> Per-shard enforcement of quotas is fine if the resource we're protecting is also per-shard (cpu time, memory), but in this instance if we're trying to protect our central controller log, I'm not sure this is going to give us as much confidence as we would like. Maybe the enforcement should be done on the controller leader instead? Or perhaps even in both places: a "friendly" check at the kafka level that does this sleep-based throttling, plus a "hard" check on the controller leader that would actually send the quota_exceeded errors if the controller log is being written to too fast.

Good points @jcsp , true that a client could jump around to other shards to end up exceeding the quota, but they could do that now anyway to sidestep the byte level (and soon to be implemented time based) quotas. Agree with your points though. 

Doing this enforcement on the leader was considered however it would be impossible to return the estimated calculated throttle response time without heavily modifying the API, which is one reason i went this route. Maybe @dotnwat can chime in here but I think the idea was to have the multi-tenancy effort really provide hard controller partition create rate quotas and KIP-599 would be our effort at providing some throttling up until we have a better solution implemented.

That being said I do agree maybe I can provide these quotas on a per-node/per-client basis, and that would be more in line with what the KIP is attempting to do.

graphcareful commented on src/v/resource_mgmt/rate.h:
```diff
@@ -99,3 +101,85 @@ class rate_tracker final {
+
+private:
+    void refill(const clock::time_point& now_ms) noexcept {
+        const auto delta = std::chrono::duration_cast<std::chrono::seconds>(
```
I did it this way because I think the algorithm necessitates that seconds are used here when refilling the number of tokens in the bucket due to the fact that the rate is expressed in terms of the window width, which is expected to be a duration in seconds.

BenPope replied:
I don't think it requires integer seconds for the calculation, though. What happens when the last check is always less than a second ago?

graphcareful replied:
So in this case, then the next window hasn't appeared so `refill()` will not add more tokens to the bucket. And if `record()` was called, the total number of tokens will then be reduced by `n`, so that checks out

graphcareful commented on src/v/kafka/server/connection_context.h:
```diff
@@ -131,6 +131,7 @@ class connection_context final
     // used to pass around some internal state
     struct session_resources {
         ss::lowres_clock::duration backpressure_delay;
+        bool delay_partition_quota_exceeded;
```
Maybe this could be cleaner? Ideally i'd like to define `quota_manager::throttle_delay` here, but that would introduce a circular dependency. Moving that class to a header such as `kafka/types.h` was considered, but I didn't want to pollute that file with such a specific class. In the end I went with this specific bool flag which is passed along to the appropriate subsystems

graphcareful commented:
Going to draft this as i'm realizing it will really depend on (KIP-482)[https://cwiki.apache.org/confluence/display/KAFKA/KIP-482%3A+The+Kafka+Protocol+should+Support+Optional+Tagged+Fields] to implement correctly. This is due to the unfortunate fact that the version bumps required in the requests that KIP-599 designates may return the new `throttling_quota_exceeded` error is greater then the introduction of the `flexibleVersion`.

We cannot support any request that has a version greater then the `flexibleVersion`. Simply attempting to would probably crash as even the parsing of the request handler will differentiate from versions older then the `flexibleVersion`.

graphcareful commented:
Rebased dev

graphcareful commented on tests/rptest/tests/cluster_quota_test.py:
```diff
@@ -0,0 +1,102 @@
+        assert bar_response[0]['ErrorCode'] == 89  # throttling_quota_exceeded
+
+        # Respect throttle millis response - exhaust the timeout so quota resets
+        throttle_ms = int(response['ThrottleMillis'])
```
don't think this int cast is necessary

dotnwat commented:
> Doing this enforcement on the leader was considered however it would be impossible to return the estimated calculated throttle response time without heavily modifying the API, which is one reason i went this route.

why is it impossible? if enforcement is done at the controller leader then it has some idea about the rate of requests.

graphcareful commented:
> > Doing this enforcement on the leader was considered however it would be impossible to return the estimated calculated throttle response time without heavily modifying the API, which is one reason i went this route.
> 
> why is it impossible? if enforcement is done at the controller leader then it has some idea about the rate of requests.

That comment is pretty old so I don't even remember what I was referring to, but doesn't it already work this way? Since `create_topics` , `delete_topics` and `create_partitions` all are executed on the controller leader, the current running totals are representative of the cluster-wide partition mutation rate.

I think that gets thrown out the window though when the controller changes, since that running total is in-memory data only stored on a single node.

dotnwat commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -54,14 +55,66 @@ throttle_delay quota_manager::record_tp_and_throttle(
+    }
+    auto it = maybe_add_and_retrieve_quota(client_id, now);
+    const auto units = it->second.pm_rate.record_and_measure(mutations, now);
+    vlog(klog.info, "UNITS: {}", units);
```
this looks like it was intended to be removed

dotnwat commented on src/v/kafka/server/handlers/create_partitions.cc:
```diff
@@ -159,6 +159,19 @@ ss::future<response_ptr> create_partitions_handler::handle(
+      valid_range_end,
+      std::back_inserter(resp.data.results),
+      error_code::invalid_request,
+      "Partition count must be greater then current number of partitions",
```
seems fine, but don't we already enforce this at the controller?

graphcareful replied:
Slightly below this commit we validate similar things such as partition count being greater then 0 and non empty assignments, i felt this belonged here with the other types of error validation

dotnwat commented on src/v/kafka/server/handlers/create_partitions.cc:
```diff
@@ -206,6 +207,29 @@ ss::future<response_ptr> create_partitions_handler::handle(
+            tp.count > cfg->partition_count,
+            "Sanity check for request increase partition count failed");
+          const auto mutations = (tp.count - cfg->partition_count);
+          const auto delay = ctx.quota_mgr().record_partition_mutations(
```
should this be broker wide enforced rather than shard local? that would probably be the second best case over enforcing rates at the controller itself. when nodes have large numbers of cores the per shard rate limit is hard to work with. per node is easier to reason about.

graphcareful replied:
done

dotnwat reviewed (commented):
lgtm. couple minor things i saw.

i wonder if we should keep do the throttling accounting on a single core so that the throttling can be expressed as per broker rather than per core. since this work is unrelated to kafka compatibility, we could even enforce these rates in the controller like john suggested. but perhaps per node is the next best thing.

mmedenjak commented:
@graphcareful @dotnwat @jcsp @mmaslankaprv @ZeDRoman how does this relate to https://github.com/redpanda-data/core-internal/issues/37? 

Overlapping? Identical? Conflicting? How much coordination do we need here?

graphcareful commented:
@dotnwat in the [KIP-599](https://cwiki.apache.org/confluence/display/KAFKA/KIP-599%3A+Throttle+Create+Topic%2C+Create+Partition+and+Delete+Topic+Operations) documentation, its called `controller_mutations_rate*`, FWIW I'm up for changing it , i'm not married to the name I've chosen

dotnwat commented:
> @dotnwat in the [KIP-599](https://cwiki.apache.org/confluence/display/KAFKA/KIP-599%3A+Throttle+Create+Topic%2C+Create+Partition+and+Delete+Topic+Operations) documentation, its called `controller_mutations_rate*`, FWIW I'm up for changing it , i'm not married to the name I've chosen

@graphcareful me neither. can you sync with @ZeDRoman about config names? i think that name is fine but maybe it ends up being name that differentiates control at the cotnroller vs control at the edge?

dotnwat commented:
Restarted CI

dotnwat commented on src/v/kafka/server/handlers/create_topics.cc:
```diff
@@ -203,6 +204,38 @@ ss::future<response_ptr> create_topics_handler::handle(
+    auto quota_exceeded_it = co_await ssx::partition(
+      begin,
+      valid_range_end,
+      [&ctx, &response](const creatable_topic& t) -> ss::future<bool> {
+          const auto delay
+            = co_await ctx.quota_mgr().record_partition_mutations(
+              ctx.header().client_id, t.num_partitions);
+          response.data.throttle_time_ms = std::max(
+            response.data.throttle_time_ms, delay);
```
looks like `response` might suffer from UAF since it is used in the coroutine lambda following the co_await?

graphcareful replied:
I have changed this method to a co-routine so `response` should be alive in this context

dotnwat replied:
the problem is not that the underlying `response` lifetime is incorrect, but that the `response` reference stored in the lambda object is destroyed and accessed after co_await. anything inside of a capture list should not be considered safe to access after a co_await in a lambda coroutine. you can copy them to the stack before co_await or sometimes it is appropriate to use ss::coroutine::lambda.

graphcareful replied:
ah thanks for the clarification, fixed

dotnwat commented on src/v/kafka/server/handlers/delete_topics.cc:
```diff
@@ -81,23 +83,57 @@ delete_topics_handler::handle(request_context ctx, ss::smp_service_group) {
+    auto quota_exceeded_it = co_await ssx::partition(
+      request.data.topic_names.begin(),
+      request.data.topic_names.end(),
+      [&ctx, &resp_delay](const model::topic& t) -> ss::future<bool> {
+          const auto cfg = ctx.metadata_cache().get_topic_cfg(
+            model::topic_namespace_view(model::kafka_namespace, t));
+          const auto mutations = cfg ? cfg->partition_count : 0;
+          const auto delay
+            = co_await ctx.quota_mgr().record_partition_mutations(
+              ctx.header().client_id, mutations);
+          resp_delay = std::max(delay, resp_delay);
```
unsafe to access resp_delay?

graphcareful replied:
same here too, `delete_topics_request::handler` is a coroutine

graphcareful replied:
fixed

dotnwat commented on src/v/kafka/protocol/schemata/delete_topics_request.json:
```diff
@@ -20,10 +20,20 @@
+  // in the response if the topics deletion is throttled (KIP-599).
+  //
+  // Version 6 reorganizes topics, adds topic IDs and allows topic names to be null.
+  "validVersions": "0-6",
   "flexibleVersions": "4+",
   "fields": [
-    { "name": "TopicNames", "type": "[]string", "versions": "0+", "entityType": "topicName",
+    { "name": "Topics", "type": "[]DeleteTopicState", "versions": "6+", "about": "The name or topic ID of the topic",
+      "fields": [
+        {"name": "Name", "type": "string", "versions": "6+", "nullableVersions": "6+", "default": "null", "entityType": "topicName", "about": "The topic name"},
+        {"name": "TopicId", "type": "uuid", "versions": "6+", "about": "The unique topic ID"}
+    ]},
+    { "name": "TopicNames", "type": "[]string", "versions": "0-5", "entityType": "topicName", "ignorable": true,
```
why are we adding version 6 if we are only bumping to version 5? i guess it doesn't affect anything, but is there something i'm missing?

graphcareful replied:
This is because I am copying the schema from upstream kafka, character for character. I don't want to make manual changes to this file. It won't matter anyway because the request version upper bound is set to 5 in `delete_topics.h`

dotnwat commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -85,26 +91,46 @@ class quota_manager {
-    // have not received any updates in ten window's worth of time.
-    void gc(clock::duration full_window);
+    // record new desired number of partition mutations
+    ss::future<std::chrono::milliseconds> record_partition_mutations(
```
since this hops cores to update usage, it probably only makes sense for low-traffic ops (like in this PR). where can we add a comment of some sort that it isn't appropriate for high-freq ops like produce or other things?

graphcareful replied:
Yeah sure, good idea

dotnwat reviewed (commented):
looks awesome. just one thing with coroutines and capture lifetime that looked problematic, but other than that, i think this is good to go.

dotnwat reviewed (commented):
See this comment: https://github.com/redpanda-data/redpanda/pull/4316#discussion_r981786825 i think i wasn't clear when I left the original review.

graphcareful commented:
In previous force push, I addressed concerns about lifetime of references on the stack within particular lambdas, and i moved the token bucket algorithm into the `kafka/` tree to not confuse others with the generic one Roman is working on that will exist in `utils/`
